libraries/standard/mqtt/src/mqtt.c

```diff
@@ -439,56 +445,87 @@ static bool matchWildcards( const char * pTopicName,
+         * current level is identified by the last character before the next level
+         * separator '/'. */
+        while( *pNameIndex < topicNameLength )
         {
-            /* Move topic name index to the end of the current level. The end of the
-             * current level is identified by '/'. */
-            while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+            /* Exit the loop if we hit the level separator. */
+            if( pTopicName[ *pNameIndex ] == '/' )
             {
-                ( *pNameIndex )++;
+                nextLevelExistsInTopicName = true;
+                break;
             }
 
-            /* Decrement the topic name index for 2 different cases:
-             * - If the break condition is ( *pNameIndex < topicNameLength ), then
-             *   we have reached past the end of the topic name and we move back the
-             *   the index on the last character.
-             * - If the break condition is ( pTopicName[ *pNameIndex ] != '/' ), we
-             *   move back the index on the '/' character. */
-            ( *pNameIndex )--;
+            ( *pNameIndex )++;
         }
 
-        /* '#' matches everything remaining in the topic name. It must be the
-         * last character in a topic filter. */
-        else if( ( pTopicFilter[ filterIndex ] == '#' ) &&
-                 ( filterIndex == ( topicFilterLength - 1U ) ) )
+        /* Determine if the topic filter contains a child level after the current level
+         * represented by the '+' wildcard. */
+        if( ( *pFilterIndex < ( topicFilterLength - 1U ) ) &&
+            ( pTopicFilter[ *pFilterIndex + 1U ] == '/' ) )
         {
-            /* Subsequent characters don't need to be checked for the
-             * multi-level wildcard. */
-            *pMatch = true;
-            shouldStopMatching = true;
+            nextLevelExistsinTopicFilter = true;
         }
-        else
+
+        /* If the topic name contains a child level but the topic filter ends at
+         * the current level, then there does not exist a match. */
+        if( ( nextLevelExistsInTopicName == true ) &&
+            ( nextLevelExistsinTopicFilter == false ) )
         {
-            /* Any character mismatch other than '+' or '#' means the topic
-             * name does not match the topic filter. */
             *pMatch = false;
             shouldStopMatching = true;
         }
+
+        /* If the topic name and topic filter have child levels, then advance the
+         * filter index to the level separator in the topic filter, so that match
+         * can be performed in the next level.
+         * Note: The name index already points to the level separator in the topic
+         * name. */
+        else if( ( nextLevelExistsInTopicName == true ) &&
+                 ( nextLevelExistsinTopicFilter == true ) )
+        {
+            ( *pFilterIndex )++;
```

Comment:
As mentioned in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1137/files#r472797473, the filter index is incremented for the case when both topic name and topic filter have child levels by the presence of the level-separator, `/`, and thus, the underflow problem for the special case is avoided.